sources/api/bootstrap-commands/src/main.rs

```diff
@@ -0,0 +1,345 @@
+
+#[derive(Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommandConfig {
+    bootstrap_commands: Option<HashMap<Identifier, BootstrapCommand>>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct BootstrapCommand {
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    commands: Option<Vec<ApiclientCommand>>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    mode: Option<BootstrapMode>,
+    #[serde(default, skip_serializing_if = "Option::is_none")]
+    essential: Option<bool>,
+}
+
+/// Stores user-supplied global arguments
+struct Args {
+    log_level: LevelFilter,
+    config_path: PathBuf,
+}
+
+impl Default for Args {
+    fn default() -> Self {
+        Self {
+            log_level: LevelFilter::Info,
+            config_path: PathBuf::from_str(DEFAULT_CONFIG_PATH).unwrap(),
+        }
+    }
+}
+
+/// Wrapper around process::Command that adds error checking.
+fn command<I, S>(bin_path: &str, args: I) -> Result<()>
+where
+    I: IntoIterator<Item = S>,
+    S: AsRef<OsStr>,
+{
+    let mut command = Command::new(bin_path);
+    command.args(args);
+    let output = command
+        .output()
+        .context(error::ExecutionFailureSnafu { command })?;
+
+    trace!("stdout: {}", String::from_utf8_lossy(&output.stdout));
+    trace!("stderr: {}", String::from_utf8_lossy(&output.stderr));
+
+    ensure!(
+        output.status.success(),
+        error::CommandFailureSnafu { bin_path, output }
+    );
+
+    Ok(())
+}
+
+/// Handles running of the api commands in a Bootstrap command
+fn handle_bootstrap_command<S>(name: S, bootstrap_command: &BootstrapCommand) -> Result<()>
+where
+    S: AsRef<str>,
+{
+    let binding = Vec::new();
+
+    let name = name.as_ref();
+    let mode = bootstrap_command.mode.clone().unwrap_or_default();
+    let essential = bootstrap_command.essential.unwrap_or(false);
+    let commands = bootstrap_command.commands.as_ref().unwrap_or(&binding);
+
+    info!("Running bootstrap command with name: {}", name);
+
+    if mode == "off" {
+        // If mode is 'off', just log this information.
+        info!("Bootstrap command mode for '{}' is 'off'", name);
+    } else {
+        let temp_path = Path::new(RUN_FILE_DIR)
+            .join(name)
+            .with_extension("ran");
+
+        // Check if bootstrap command is already run in this boot.
+        if temp_path.exists() {
+            info!("Skipping Bootstrap command '{}' because its already run", name);
+            return Ok(());
+        }
+
+        // Create a file in /run/bootstrap-commands to indicate an attempt to run this bootstrap-command
+        command(constants::TOUCH_BIN, [temp_path])?;
+
+        info!("Processing Bootstrap command '{}' ... ", name);
+        for cmd in commands.iter() {
+            // this condition is already taken care of
+            let args = cmd.get_args();
```

Comment:
You can do:

```
let args = cmd.get_args().context(error::BootstrapCommandFailureSnafu { name })?;
```

And avoid the `match` arms, which is also incorrect since the way it is written doesn't actually trigger an error.

Reply:
The discussed `get_args` has a default value which makes the function infallible which is why I decided not to use snafu in this case.